benchmark/metrics_parser.py

```diff
@@ -0,0 +1,276 @@
+    return mean(values)
+
+
+def wallclock_for_operation(results, operation):
+    """
+    Calculate the wallclock time for a process running in a particular
+    scenario.
+
+    By default this will return the `mean` result.
+    """
+    operation_results = itertools.ifilter(
+        lambda r: r['metric']['type'] == 'wallclock'
+        and r['operation']['type'] == operation,
+        results
+    )
+    values = [r['value'] for r in operation_results]
+    return mean(values)
+
+
+def container_convergence(results, seconds):
+    """
+    Calculate the percentage of containers that converge within a given
+    time period.
+    """
+    convergence_results = [
+        r for r in results if r['metric']['type'] == 'wallclock'
+        and r['operation']['type'] == 'create-container'
+    ]
+    num_convergences = len(convergence_results)
+    if num_convergences > 0:
+        convergences_within_limit = [
+            r for r in convergence_results if r['value'] < seconds
+        ]
+        return len(convergences_within_limit) / num_convergences
+
+    return None
+
+
+def request_latency(results, seconds):
+    """
+    Calculate the percentage of scenario requests have a latency under the
+    specified time limit.
+    """
+    scenario_results = [
+        r['scenario'] for r in results
+        if r['scenario'].get('metrics')
+        and r['scenario']['metrics'].get('call_durations')
+    ]
+
+    if len(scenario_results) > 0:
+        unique_metrics = []
+        for result in scenario_results:
+            if result['metrics'] not in unique_metrics:
+                unique_metrics.append(result['metrics'])
+
+        total_requests = 0
+        requests_under_limit = 0
+        for metric in unique_metrics:
+            for k, v in metric['call_durations'].iteritems():
+                if float(k) < seconds:
+                    requests_under_limit += v
+            total_requests += metric['ok_count'] + metric['err_count']
+        return requests_under_limit / total_requests
+    return None
+
+
+class BenchmarkingResults(object):
+    """
+    Processes benchmarking results and produces reports.
+    """
+    def __init__(self):
+        self.results = []
+
+    def add_results(self, results):
+        """
+        Add a set of results to the existing results.
+        """
+        self.results.extend(self._flatten(results))
+
+    def output_csv(self, prefix):
+        """
+        Output a CSV representation of a set of results.
+        """
+        summary = self._create_summary()
+        for scenario, result in summary.iteritems():
+            write_csv(result, '{0}-{1}.csv'.format(prefix, scenario))
+
+    def _versions(self):
+        """
+        Return all unique versions present in the results.
+        """
+        versions = [
+            r['control_service']['flocker_version'] for r in self.results
+        ]
+        return list(set(versions))
```

Comment:
It's okay for these to be sets as itertools just needs iterables, not lists. So you can just do:

return set(<list comprehension code>)